src/plugins/intel_cpu/src/emitters/snippets/x64/jit_loop_emitters.cpp

```diff
@@ -48,22 +58,25 @@ void jit_loop_begin_emitter::validate_arguments(const std::vector<size_t> &in, c
 void jit_loop_begin_emitter::emit_code(const std::vector<size_t> &in, const std::vector<size_t> &out,
                                        const std::vector<size_t> &pool_vec_idxs, const std::vector<size_t> &pool_gpr_idxs) const {
     validate_arguments(in, out);
-    jit_emitter::emit_code(in, out, pool_vec_idxs, pool_gpr_idxs);
-}
 
-void jit_loop_begin_emitter::emit_impl(const std::vector<size_t>& in, const std::vector<size_t>& out) const {
     // If the loop evaulate once, we can skip loop begin code emission
     // If work_amount is dynamic, we should get runtime `work_amount` - it might be `zero` and we should skip loop evaluation
     if (evaluate_once && !is_work_amount_dynamic)
         return;
 
     Reg64 reg_work_amount = Reg64(static_cast<int>(out.back()));
     if (is_work_amount_dynamic) {
-        Reg64 reg_runtime_params = abi_param1;  // defined by jit_kernel_emitter
-        Reg64 reg_loop_args_ptr = Reg64(static_cast<int>(aux_gpr_idxs[0]));
+        Reg64 reg_loop_args_ptr;
+        const auto preserved = init_aux_gpr(reg_loop_args_ptr, pool_gpr_idxs, out);  // loop_begin doesn't have input registers
+        if (preserved)
+            h->push(reg_loop_args_ptr);
+
         const auto id_offset = loop_id * sizeof(jit_snippets_call_args::loop_args_t);
-        h->mov(reg_loop_args_ptr, h->ptr[reg_runtime_params + GET_OFF(loop_args)]);
+        h->mov(reg_loop_args_ptr, h->ptr[abi_param1 + GET_OFF(loop_args)]); // defined by jit_kernel_emitter
         h->mov(reg_work_amount, h->ptr[reg_loop_args_ptr + id_offset + GET_OFF_LOOP_ARGS(m_work_amount)]);
+
+        if (preserved)
+            h->pop(reg_loop_args_ptr);
```

Comment:
This would work, but it's still an error-prone design because we need to push/pop manually. Sooner or later we (or smone else) would make another mistake with that.
So why don't we do it right straight away? 🙂 
Let's create a class `aux_reg_holder` that would `push`/take from the pool in constructor and `pop`/return to the pool in destructor? It would have only one public method: const Reg64& get_reg();
This way we'll guarantee that we never forget to restore the register/reg pool state, when the tmp reg is not needed anymore. It would also simplify our emitters, so we won't have to write push/pop manually
What do you think?

Reply:
The only obvious drawback of this approach is that sometimes (rarely) we might need to destroy the instance when there are no obvious scope boundaries (e.g. `for`/`if`). The easiest and the best way to solve it is to limit the scope manually({...}), another option is to create `release()` method that would be normally called from the destructor, but can be called manually as well. We would also need a check that `get_reg` is called only on non-released state. But I would rather drop this functionality completely to keep it simple and used {} instead.